gamemodes/terrortown/gamemode/client/cl_help.lua

```diff
@@ -339,71 +336,27 @@ function HELPSCRN:ShowSubmenu(menuClass)
 
 	-- BUILD GENERAL BOX STRUCTURE
 	local navArea = vgui.Create("DNavPanelTTT2", frame)
-	navArea:SetSize(widthNav, heightNav - vskin.GetHeaderHeight() - vskin.GetBorderSize())
+	navArea:SetWide(widthNav)
 	navArea:SetPos(0, 0)
+	navArea:DockPadding(0, self.padding, 1, self.padding)
 	navArea:Dock(LEFT)
 
-	local navAreaContent = vgui.Create("DPanel", navArea)
-	navAreaContent:SetPos(0, heightNavHeader)
-	navAreaContent:SetSize(widthNavContent, heightNavContent - vskin.GetHeaderHeight() - vskin.GetBorderSize())
-
-	-- MAKE NAV AREA SCROLLABLE
-	local navAreaScroll = vgui.Create("DScrollPanelTTT2", navAreaContent)
-	navAreaScroll:SetVerticalScrollbarEnabled(true)
-	navAreaScroll:Dock(FILL)
-
-	-- SPLIT NAV AREA INTO A GRID LAYOUT
-	local navAreaScrollGrid = vgui.Create("DIconLayout", navAreaScroll)
-	navAreaScrollGrid:Dock(FILL)
-	navAreaScrollGrid:SetSpaceY(self.padding)
-
 	local contentArea = vgui.Create("DContentPanelTTT2", frame)
 	contentArea:SetSize(widthContent, heightContent - vskin.GetHeaderHeight() - vskin.GetBorderSize())
 	contentArea:SetPos(widthNav, 0)
 	contentArea:DockPadding(self.padding, self.padding, self.padding, self.padding)
 	contentArea:Dock(TOP)
 
-	-- GENERATE MENU CONTENT
-	local subMenuClasses = menuClass:GetVisibleSubmenus()
-
-	-- cache reference to last active button
-	local lastActive
-
-	if #subMenuClasses == 0 then
-		local labelNoContent = vgui.Create("DLabelTTT2", contentArea)
-		labelNoContent:SetText("label_menu_not_populated")
-		labelNoContent:SetSize(widthContent - 40, 50)
-		labelNoContent:SetFont("DermaTTT2Title")
-		labelNoContent:SetPos(20, 0)
-	else
-		for i = 1, #subMenuClasses do
-			local subMenuClass = subMenuClasses[i]
-
-			local settingsButton = navAreaScrollGrid:Add("DSubmenuButtonTTT2")
-			settingsButton:SetSize(widthNavButton, heightNavButton)
-			settingsButton:SetTitle(subMenuClass.title or subMenuClass.type)
-			settingsButton:SetIcon(subMenuClass.icon)
-
-			settingsButton.DoClick = function(slf)
-				HELPSCRN:SetupContentArea(contentArea, subMenuClasses[i])
-				HELPSCRN:BuildContentArea()
+	-- MAKE SEPARATE SUBMENULIST ON THE NAVAREA WITH A CONTENT AREA
+	local submenuList = vgui.Create("DSubmenuListTTT2", navArea)
+	submenuList:Dock(FILL)
+	submenuList:SetPadding(self.padding)
+	submenuList:SetSubmenuClasses(menuClass:GetVisibleSubmenus(), contentArea)
+	submenuList:EnableSearchBar(menuClass:HasSearchbar())
+	submenuList:SetSearchFunction(menuClass:GetSearchFunction())
```

Comment:
Here I just want to store it. But yeah, as I said, I can just write menuClass.SearchFunction instead

Reply:
I see what you mean. Should the search function be really defined like this? I mean the search is always dependant on the menu you are in. So why do you need to link this function anyways?

Reply:
Mainly because I access it in a for-loop, so it's called very often. Localizing it should make it more efficient as far as I understood your performance improvements on older code.

Reply:
Okay, hear me out. Take this part of the code:

https://github.com/TTT-2/TTT2/pull/835/files#diff-b5666cfb68877262c5ef28ff03124e5a00e32b08ced4be0a298615727238c6b2R86-R94

And move it to the menu search code. Then you only have to make one function call. Would this work?

Reply:
Something on the lines of:

```lua
function CLGAMEMODEMENU:GetMatchingSubmenus(searchString)
	-- iterate over all submenus and return a list of matching menus
end
```

This would also fit the style of the submenu lists and virtual menus

Reply:
I see your reasoning that this would create code duplication. You could make the search string matching a secondary function on `CLGAMEMODEMENU`. In normal cases only the matching function gets overwritten. But if one wants to, they can also overwrite the whole search system.

Reply:
I thought about this, but it gives to much freedom to the user. And everytime I want to make a custom searchfunction, I have to copy a lot of code and look out for edge-cases. I commited the better version, because you are right with the complicated style I used.

Reply:
Yeah that might allow to rewrite it, but I thought about making the search way more effective with caching for long lists and searches. I rather want to keep the outer logic to TTT2 and only allow the user to define when a search results in true and when not ^^

Reply:
Is too much freedom to the developer a bad thing? And if you have a `CLGAMEMODEMENU:GetMatchingSubmenus` and a `CLGAMEMODEMENU:MatchSearchString` function where you overwrite only the latter, would this really be more copy and paste then your initial one? Because the matching function basically is exactly your initial function, you just moved the search iteration code from the menu element to the menu class. Which is not a bad thing imho, because I think the menu logic should be contained in the menu class.

Reply:
> Yeah that might allow to rewrite it, but I thought about making the search way more effective with caching for long lists and searches. I rather want to keep the outer logic to TTT2 and only allow the user to define when a search results in true and when not ^^

That's okay with me. But I don't see what you can cache in the current version, that you can't cache in the menu-class-version

Reply:
Looking at your most recent commit I see that you have the matching function like I suggested, only the search function is not in the class. I'm fine either way and I think I voiced my opinion on this enough. Let's see what the rest of the team has to say about this.

Reply:
Your last point is also valid. I see how I could easily make the same performance-optimization like this. Then you can only change what you want. I still have my simple matchsearchstring function, where I don't have to rewrite the outer loop and anybody could change it anyways :D

Reply:
I know the feeling of someone critisizing the code that one thinks is so brilliant. Maybe give this problem a day to think about and then decide if you like your solution or my suggestion more.
I often noticed that I hated such suggestions in the beginning but came around liking them in the end after thinking about it for a while

Reply:
Nah, thats not my problem ^^
I like your suggestions so far, you convinced me quite fast xD

It's just really difficult to see clear after a few days of deep coding. I already refactored a lot and am always surprised of what I didnt think of. The code on the base_gamemodemenu was the first I did and then didnt touch anymore, because it just worked :stuck_out_tongue_closed_eyes:

Reply:
I really appreciate it, that you care about the code and look out for the style, many other people would just ask:
"Does it work?"
"Yes"
"Okay, merge it!" 😄

Reply:
I learned from the best @Alf21 and @saibotk who always take my PRs apart, lol :D